eax/eax.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019 ProtonTech AG
+
+// Package eax provides an implementation of the EAX
+// (encrypt-authenticate-translate) mode of operation, as described in
+// Bellare, Rogaway, and Wagner "THE EAX MODE OF OPERATION: A TWO-PASS
+// AUTHENTICATED-ENCRYPTION SCHEME OPTIMIZED FOR SIMPLICITY AND EFFICIENCY."
+// In FSE'04, volume 3017 of LNCS, 2004
+
+package eax
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"errors"
+	"golang.org/x/crypto/internal/byteutil"
+)
+
+const (
+	defaultTagSize   = 16
+	defaultNonceSize = 16
+)
+
+type eax struct {
+	block     cipher.Block // Only AES-{128, 192, 256} supported
+	tagSize   int          // At least 12 bytes recommended
+	nonceSize int
+}
+
+func (e *eax) NonceSize() int {
+	return e.nonceSize
+}
+
+func (e *eax) Overhead() int {
+	return e.tagSize
+}
+
+// NewEAX returns an EAX instance with AES-{keyLength} and default tag and
+// nonce lengths.
+func NewEAX(key []byte) (cipher.AEAD, error) {
+	return NewEAXWithNonceAndTagSize(key, defaultNonceSize, defaultTagSize)
+}
+
+// NewEAXWithNonceAndTagSize returns an EAX instance with AES-{keyLength} and
+// given nonce and tag lengths in bytes. Panics on zero nonceSize and
+// exceedingly long tags.
+//
+// It is recommended to use at least 12 bytes as tag length (see, for instance,
+// NIST SP 800-38D).
+//
+// Only to be used for compatibility with existing cryptosystems with
+// non-standard parameters. For all other cases, prefer NewEAX.
+func NewEAXWithNonceAndTagSize(
+	key []byte, nonceSize, tagSize int) (cipher.AEAD, error) {
+	if nonceSize < 1 {
+		return nil, eaxError("Cannot initialize EAX with nonceSize = 0")
+	}
+	aesCipher, err := aes.NewCipher(key)
+	if err != nil {
+		panic(err)
+	}
+	if tagSize > aesCipher.BlockSize() {
+		return nil, eaxError("Custom tag length exceeds blocksize")
+	}
+	return &eax{
+		block:     aesCipher,
+		tagSize:   tagSize,
+		nonceSize: nonceSize,
+	}, nil
+}
+
+// Seal (AEAD interface) returns a byte array containing the
+// concatenation of the ciphertext and the validation tag.
+func (e *eax) Seal(dst, nonce, plaintext, adata []byte) []byte {
```

Comment:
The AEAD interface claims that Seal and Open append the result to `dst`, but it is unused here. (See `sliceForAppend` at https://golang.org/src/crypto/cipher/gcm.go#L170.)

Reply:
Thanks for pointing this out. Crucial for memory management.